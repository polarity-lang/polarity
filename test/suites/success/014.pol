data Nat { Z, S(n: Nat) }

def Nat.add(y: Nat) : Nat {
    Z => y,
    S(x') => S(x'.add(y)),
}

data Bool { True, False }

data Vec(n: Nat) {
    Nil: Vec(Z),
    Cons(n: Nat, x: Bool, xs: Vec(n)): Vec(S(n))
}

def Vec(n).append'(n m: Nat, ys: Vec(m)): Vec(n.add(m)) {
    Cons(n', x, xs) => Cons(n'.add(m), x, xs.append'(n', m, ys)),
    Nil => ys
}

let append(n m: Nat, xs: Vec(n), ys: Vec(m)): Vec(n.add(m)) {
    xs.append'(n, m, ys)
}
