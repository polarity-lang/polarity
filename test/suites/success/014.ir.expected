def .add(y) {
    Z => y,
    S(x') => S(x'.add(y)),
}

def .append'(n, m, ys) {
    Cons(n', x, xs) => Cons(n'.add(m), x, xs.append'(n', m, ys)),
    Nil => ys,
}

let append(n, m, xs, ys) { xs.append'(n, m, ys) }
